Prefer unwrap_or_else to unwrap_or in case of function calls/allocations

ljedrz opened the pull request:
The contents of `unwrap_or` are evaluated eagerly, so it's not a good pick in case of function calls and allocations. This PR also changes a few `unwrap_or`s with `unwrap_or_default`.

An added bonus is that in some cases this change also reveals if the object it's called on is an `Option` or a `Result` (based on whether the closure takes an argument).

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

Xanewok commented on src/bootstrap/config.rs:
```diff
@@ -430,7 +430,7 @@ impl Config {
             }
         }).unwrap_or_else(|| TomlConfig::default());
 
-        let build = toml.build.clone().unwrap_or(Build::default());
```
I think here and in many places further down this could be replaced with lazy `unwrap_or_default()`, which may even further reduce verbosity (e.g. `unwrap_or_else(|| (String::new(), String::new()))`)

ljedrz replied:
Definitely; I'll update the commit soon.

killercup commented on src/bootstrap/compile.rs:
```diff
@@ -544,7 +544,7 @@ pub fn rustc_cargo_env(builder: &Builder, cargo: &mut Command) {
          .env("CFG_CODEGEN_BACKENDS_DIR", &builder.config.rust_codegen_backends_dir);
 
-    let libdir_relative = builder.config.libdir_relative().unwrap_or(Path::new("lib"));
+    let libdir_relative = builder.config.libdir_relative().unwrap_or_else(|| Path::new("lib"));
```
Hm, ideally, using the `_else` variant could be a great indicator whether the computation is actually expensive. This change, for example, is not necessary: It [is](https://doc.rust-lang.org/1.29.2/src/std/path.rs.html#1698-1700) just a cast.

I'm fine with changing all function calls to `_else` for now, though.

rkruppe replied:
Yeah most of these `_else` are line noise with no positive impact. Many wrap functions like `String::new`, `Vec::new`, tuple struct constructors, etc. that are literally free.

ljedrz replied:
@rkruppe well, most are `to_string()`, which are not free.

rkruppe replied:
Fair, I haven't actually counted, but I did see a significant number of them.

ljedrz replied:
@rkruppe you are 100% right about `Vec/String::new`, though; I knew they didn't allocate, but I didn't notice they had become `const fn`.

ljedrz commented:
Comments addressed; I replaced calls to `unwrap_or_else(|| Vec/String::new())` (and a few applicable others) with `unwrap_or_default()`, I also removed the accidental change to one tuple struct constructor.

bors commented:
:umbrella: The latest upstream changes (presumably #54941) made this pull request unmergeable. Please resolve the merge conflicts.

ljedrz commented:
@matthewjasper rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #55171) made this pull request unmergeable. Please resolve the merge conflicts.

ljedrz commented:
Rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #54976) made this pull request unmergeable. Please resolve the merge conflicts.

ljedrz commented:
Re(eee)based. @kennytm these are trivial changes good for a rollup if they can be squeezed in; I won't be able to rebase for the next few days.

matthewjasper commented:
@bors r+

bors commented:
:pushpin: Commit d28aed6dc45ffccc790469cb04f3f775ddb2283a has been approved by `matthewjasper`

<!-- @bors r=matthewjasper d28aed6dc45ffccc790469cb04f3f775ddb2283a -->

bors commented:
:hourglass: Testing commit d28aed6dc45ffccc790469cb04f3f775ddb2283a with merge 398c664f3a4b3cfb83992c05604f87266a029e38...

matthewjasper commented:
@bors retry

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19658664), [status-travis](https://travis-ci.org/rust-lang/rust/builds/444052198?utm_source=github_status&utm_medium=notification)
Approved by: matthewjasper
Pushing ca2639e82ec4a18d7359efbfb555ea69dd644c97 to master...
